File: Week_03-递归-分治-回溯/169.多数元素.py
#
# @lc app=leetcode.cn id=169 lang=python
#
# [169] 多数元素
#

# @lc code=start
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        '''
        # 方法1:哈希表方法
        # 建立哈希表时，同步进行打擂台，维护出现次数最多的元素及其出现次数
        dictable = dict()
        cnt = 0
        for i in range(len(nums)):
            if nums[i] in dictable:
                dictable[nums[i]] += 1
            else:
                dictable[nums[i]] = 1
            if dictable[nums[i]] > cnt:
                majority = nums[i]
                cnt += 1
        return majority
        '''
        #方法2，候选多数元素打擂台
        #假定当前元素是候选多数元素，遍历后续元素并对其计数
        #如果和候选元素相同，计数+1
        #如果和候选元素不同，技术减1，如果减到了0，说明其可能不是候选多数元素，则把当前元素列为候选多数元素，从1开始对其进行计数
        #遍历结束，最后的候选元素即为结果
        target = nums[0]
        cnt = 1
        for i in range(1, len(nums)):
            if nums[i] != target:
                cnt -= 1
                if cnt == 0:
                    target = nums[i]
                    cnt = 1
            else:
                cnt += 1
        return target

        
# @lc code=end

File: Week_03-递归-分治-回溯/169.多数元素.cpp
/*
 * @lc app=leetcode.cn id=169 lang=cpp
 *
 * [169] 多数元素
 */
// @lc code=start
class Solution {
public:
    int majorityElement(vector<int>& nums) {
#if 0
/*
哈希表存储各个元素出现次数；
方法一：遍历哈希表，看哪个元素出现次数多；
方法二：建立哈希表时，同步进行打擂台，维护出现次数最多的元素及其出现次数
*/       
        int majority;
        unordered_map<int,int> um;
        int cnt = 0;
        for (auto n : nums) {
            um[n]++;
            if (um[n] >cnt) {
                majority = n;
                cnt = um[n];
            }
        }
        return majority;
#endif
#if 1
/*
原理，假定第一个元素是多数元素，遍历后续所有元素，出现和当前假定多数元素相同的元素，+1；不同的元素，-1；
如果当前假定元素的计数等于0，则其可能不是多数元素，变更假定多少元素为当前遍历到的元素
遍历结束，最后的假定多数元素是真的多数元素
*/
        int target = nums[0];
        int count = 1;
        for(int i = 1;i<nums.size();i++){
            if(nums[i]!=target) {  
                count--;
                if(count==0)
                {
                    target = nums[i];
                    count = 1;
                }
            } else count++;
        }
        return target;
#endif
    }

};
// @lc code=end

